### PR TITLE
Improve performance of trace page

### PR DESCRIPTION
**Pull request by tacigar**
* Calculate Trace summaries asynchronously.
* Reuse results of `/api/v2/traces` API call in the individual trace page if can.

### CODE REVIEW

**Comment by tacigar** [comment]
I'll update unit tests.

**Comment by codefromthecrypt** [comment]
as I understand, this improves performance by doing work in parallel correct?

File: zipkin-lens/src/reducers/trace.js [hunk 1]
```diff
@@ -15,28 +15,28 @@ import * as types from '../constants/action-types';
...
 
 const trace = (state = initialState, action) => {
   switch (action.type) {
-    case types.FETCH_TRACE_REQUEST:
+    case types.TRACE_LOAD_REQUEST:
       return {
         ...state,
         isLoading: true,
-        trace: [], /* Initialize trace */
+        traceSummary: {}, /* Initialize trace */
```
Comment (zeagord) [inline zipkin-lens/src/reducers/trace.js]: If it available in the initialstate, do we need still initialize agian?

**Review by zeagord — CHANGES_REQUESTED** [review]
Looks good to me

**Comment by tacigar** [comment]
@adriancole 
Yes, correct.
And this improves performance by avoiding calling `/api/v2/trace` API in trace page if not needed.
In other words, reuse results of `/api/v2/traces` API in the individual trace page.

File: zipkin-lens/src/reducers/trace.js (hunk 1 above)
Comment (tacigar) [reply zipkin-lens/src/reducers/trace.js]: Yes, I think so.
I think previous `traceSummary` state should be cleaned up before loading a new trace.

**Review by zeagord — APPROVED** [review]

**Comment by tacigar** [comment]
PTAL 🙇

File: zipkin-lens/src/actions/trace-action.js [hunk 2]
```diff
@@ -13,36 +13,50 @@
...
+const calculateCorrectedTrace = async trace => treeCorrectedForClockSkew(trace);
+
+const calculateDetailedTraceSummary = async correctedTrace => buildDetailedTraceSummary(
+  correctedTrace,
+);
+
+export const loadTrace = (traceId, correctedTraceMap) => async (dispatch) => {
+  dispatch(loadTraceRequest());
+
+  if (correctedTraceMap[traceId]) {
```
Comment (tacigar) [inline zipkin-lens/src/actions/trace-action.js:43]: Reuse `correctedTrace` of `correctedTraceMap` which is created in `loadTraces` action creator.

File: zipkin-lens/src/actions/trace-action.js [hunk 3]
```diff
@@ -13,36 +13,50 @@
...
+  } else {
+    try {
+      const res = await fetch(`${api.TRACE}/${traceId}`);
+
+      if (!res.ok) {
+        throw Error(res.statusText);
+      }
+      const trace = await res.json();
+      const correctedTrace = await calculateCorrectedTrace(trace);
+      const detailedTraceSummary = await calculateDetailedTraceSummary(correctedTrace);
```
Comment (tacigar) [inline zipkin-lens/src/actions/trace-action.js]: Calculate detailedTraceSummary in action creator.

File: zipkin-lens/src/containers/TracePage/TracePageContainer.js [hunk 4]
```diff
@@ -13,31 +13,31 @@
...
 
 const mapDispatchToProps = dispatch => ({
-  fetchTrace: traceId => dispatch(fetchTrace(traceId)),
+  loadTrace: (traceId, correctedTraceMap) => dispatch(loadTrace(traceId, correctedTraceMap)),
 });
 
 const TracePageContainer = connect(
   mapStateToProps,
   mapDispatchToProps,
+  (stateProps, dispatchProps, ownProps) => ({
```
Comment (tacigar) [inline zipkin-lens/src/containers/TracePage/TracePageContainer.js:35]: This argument is **mergeProps**.

https://github.com/reduxjs/react-redux/blob/master/docs/api/connect.md#mergeprops-stateprops-dispatchprops-ownprops--object

Comment (codefromthecrypt) [reply zipkin-lens/src/containers/TracePage/TracePageContainer.js:35]: this could make a good code comment

**Comment by tacigar** [comment]
PTAL

**Comment by codefromthecrypt** [comment]
thanks much!

**Comment by shakuzen** [comment]
> And this improves performance by avoiding calling `/api/v2/trace` API in trace page if not needed.
> In other words, reuse results of `/api/v2/traces` API in the individual trace page.

I haven't thoroughly checked the implementation (sorry), but I'm wondering if a user wants to get the latest data from `/api/v2/trace` on the trace view screen, is there a way to do this? Does refreshing the web page do that? I can imagine in the case that the trace is in-progress when returned in the search results, a user will want a way to get the up-to-date trace data in the trace detail view.

**Pull request merged**